pkg/istio-agent/xds-proxy.go

```diff
@@ -141,17 +153,42 @@ func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDisc
 			// todo close downstream?
 			return err
 		case req := <-requestsChan:
+			if req.TypeUrl == v3.ListenerType {
+				// for every LDS req, add a NDS req
+				requestsChan <- &discovery.DiscoveryRequest{
+					// Rather than inventing our own nonces etc.
+					// just copy the same nonces/version info from lds
+					// i.e. if lds fails, nds is assumed to have failed
+					VersionInfo:   req.VersionInfo,
+					Node:          req.Node,
+					ResourceNames: nil,
+					TypeUrl:       v3.NameTableType,
+					ResponseNonce: req.ResponseNonce,
+				}
+			}
 			if err = upstream.Send(req); err != nil {
 				proxyLog.Errorf("upstream send error: %v", err)
 				return err
 			}
 		case resp := <-responsesChan:
-			if err := downstream.Send(resp); err != nil {
-				proxyLog.Errorf("downstream send error: %v", err)
-				// we cannot return partial error and hope to restart just the downstream
-				// as we are blindly proxying req/responses. For now, the best course of action
-				// is to terminate upstream connection as well and restart afresh.
-				return err
+			if resp.TypeUrl == v3.NameTableType {
+				// intercept. This is for the dns server
+				if p.localDNSServer != nil {
+					var nt nds.NameTable
+					if err = ptypes.UnmarshalAny(resp.Resources[0], &nt); err != nil {
```

Comment:
`resp.Resources[0]` assumes we will get a valid response. Probably not safe to assume, as we can crash here on a bad response